### ISSUE TITLE
Possible lack of negative filtering

### ISSUE THREAD

**Issue by zepalmer** [JHU-PL-Lab/odefa#2]
Now that the implementation is actually reporting filtered values as opposed to unfiltered values, there may be some issue with the application of negative filters.   In [1ddpa_conditional_doesnt_use_context_stack.code](../blob/master/test-sources/1ddpa_conditional_doesnt_use_context_stack.code), for instance, the listed expectation for a value which flows through the antimatch branch doesn't include any filters (and the unit tests pass).  Doesn't this indicate a lack of proper negative filtering by the implementation?

**Comment by leafac** [comment]
If I understand you right, you're claiming that the linked test should have the following in place of line 5:

```
# EXPECT-ANALYSIS-LOOKUP-FROM-END ib {{b=e$}:(+{},-{{(a, {})}})}
```

to which I disagree.

When we do a positive filter—for example, filtering the record `{a = e}` with the filter `{a: {}}`—we need to keep (at least a part of) the filter around to validate the inner values—in the example, validate that whatever shows up at `e` matches the `{}` pattern.

On the other hand, when we find that `{b = e}` _doesn't_ match `{a: {}}`, we can discard the `{a: {}}` filter at once. We don't need to keep it around for further validation, because the work is over, so we might as well just drop it.

If we do that, then the test is fine as it is.

**Comment by zepalmer** [comment]
I agree with your assertion: it's not strictly necessary to keep this information around.  That said, what we are describing is an optimization and I don't remember including such an optimization when we reasoned about the rules.  I'm a bit concerned that we might not be handling negative filters properly _at all_ and this is a symptom of that.  I created a bug report 'cause I don't have the time to design and debug a proper test case.

Thanks for taking a look; it's good to get some external perspective.  I think the resulting rule set is actually a whole lot nicer as a result of the filtered value changes; I just want to make sure I haven't broken anything.
